Wrong gateway is detected starting with Testcontainers 1.12.4

Hi,

I have a jenkins CI with a master node and two slaves. All tests are running smoothly on slaves but they fail once run on master. Master and slaves differs in terms of ubuntu version and docker version.
The master is:
```Server Version: 19.03.4
API Version: 1.40
Operating System: Ubuntu 18.04.3 LTS
```
While the slaves are:
```
Server Version: 19.03.0
API Version: 1.40
Operating System: Ubuntu 16.04.6 LTS
```

I use testcontainers 1.12.4

Exceptions I am getting on master node:
```
13:11:08  [info]   org.testcontainers.containers.ContainerLaunchException: Container startup failed
13:11:08  [info]   at org.testcontainers.containers.GenericContainer.doStart(GenericContainer.java:322)
13:11:08  [info]   at org.testcontainers.containers.GenericContainer.start(GenericContainer.java:302)
13:11:08  [info]   at com.dimafeng.testcontainers.SingleContainer.start(Container.scala:42)
13:11:08  [info]   ...
13:11:08  [info]   Cause: org.testcontainers.containers.ContainerFetchException: Can't get Docker image: RemoteDockerImage(imageNameFuture=java.util.concurrent.CompletableFuture@687cf499[Completed normally], imagePullPolicy=DefaultPullPolicy(), dockerClient=LazyDockerClient.INSTANCE)
13:11:08  [info]   at org.testcontainers.containers.GenericContainer.getDockerImageName(GenericContainer.java:1265)
13:11:08  [info]   at org.testcontainers.containers.GenericContainer.logger(GenericContainer.java:600)
13:11:08  [info]   at org.testcontainers.containers.GenericContainer.doStart(GenericContainer.java:311)
13:11:08  [info]   at org.testcontainers.containers.GenericContainer.start(GenericContainer.java:302)
13:11:08  [info]   at com.dimafeng.testcontainers.SingleContainer.start(Container.scala:42)
13:11:08  [info]   ...
13:11:08  [info]   Cause: java.lang.IllegalStateException: Can not connect to Ryuk
13:11:08  [info]   at org.testcontainers.utility.ResourceReaper.start(ResourceReaper.java:150)
13:11:08  [info]   at org.testcontainers.DockerClientFactory.client(DockerClientFactory.java:135)
13:11:08  [info]   at org.testcontainers.LazyDockerClient.getDockerClient(LazyDockerClient.java:14)
13:11:08  [info]   at org.testcontainers.LazyDockerClient.listImagesCmd(LazyDockerClient.java:12)
13:11:08  [info]   at org.testcontainers.images.LocalImagesCache.maybeInitCache(LocalImagesCache.java:68)
13:11:08  [info]   at org.testcontainers.images.LocalImagesCache.get(LocalImagesCache.java:32)
13:11:08  [info]   at org.testcontainers.images.AbstractImagePullPolicy.shouldPull(AbstractImagePullPolicy.java:18)
13:11:08  [info]   at org.testcontainers.images.RemoteDockerImage.resolve(RemoteDockerImage.java:62)
13:11:08  [info]   at org.testcontainers.images.RemoteDockerImage.resolve(RemoteDockerImage.java:25)
13:11:08  [info]   at org.testcontainers.utility.LazyFuture.getResolvedValue(LazyFuture.java:20)
```

and
```
12:11:37.140 [testcontainers-ryuk] WARN  o.t.u.ResourceReaper - Can not connect to Ryuk at localhost:32965
13:11:37  java.net.SocketException: Connection reset
13:11:37  	at java.net.SocketInputStream.read(SocketInputStream.java:210)
13:11:37  	at java.net.SocketInputStream.read(SocketInputStream.java:141)
13:11:37  	at sun.nio.cs.StreamDecoder.readBytes(StreamDecoder.java:284)
13:11:37  	at sun.nio.cs.StreamDecoder.implRead(StreamDecoder.java:326)
13:11:37  	at sun.nio.cs.StreamDecoder.read(StreamDecoder.java:178)
13:11:37  	at java.io.InputStreamReader.read(InputStreamReader.java:184)
13:11:37  	at java.io.BufferedReader.fill(BufferedReader.java:161)
13:11:37  	at java.io.BufferedReader.readLine(BufferedReader.java:324)
13:11:37  	at java.io.BufferedReader.readLine(BufferedReader.java:389)
13:11:37  	at org.testcontainers.utility.ResourceReaper$FilterRegistry.waitForAcknowledgment(ResourceReaper.java:406)
13:11:37  	at org.testcontainers.utility.ResourceReaper$FilterRegistry.register(ResourceReaper.java:402)
13:11:37  	at org.testcontainers.utility.ResourceReaper.lambda$start$1(ResourceReaper.java:128)
13:11:37  	at java.lang.Thread.run(Thread.java:748)
```

Although when I call docker logs on master node 
```
+ docker ps
CONTAINER ID        IMAGE                               COMMAND             CREATED             STATUS              PORTS                     NAMES
7b115c90863a        quay.io/testcontainers/ryuk:0.2.3   "/app"              32 seconds ago      Up 31 seconds       0.0.0.0:32965->8080/tcp   testcontainers-ryuk-75c3e431-6dd5-4faf-9b54-df86d395ff4e
++ docker ps -q
+ docker logs 7b115c90863a
2020/01/12 10:08:16 Starting on port 8080...
```
I can see that ryuk has started.

I assume that this is not a fault of ubuntu neither of docker version (mainly because I use similar versions on my local machine and it all works), but rather some specific configuration of master node. Unfortunately I have no idea how to debug it further, so any help would be appreciated.

Cheers!

Changed the code a little bit since `DockerClientProviderStrategy` is not in 1.12.4, so I used `EnvironmentAndSystemPropertyClientProviderStrategy` instead.

```
        val client = DockerClientFactory.instance().client()
        val network = client.inspectNetworkCmd()
            .withNetworkId("bridge")
            .exec();
        println("Network: $network");

        val gateway = network
            .getIpam()
            .getConfig()
            .stream()
            .filter { it -> it.getGateway() != null }
            .findAny()
            .map { it.gateway }
            .orElse("localhost");

        System.out.println("Gateway: $gateway")

        EnvironmentAndSystemPropertyClientProviderStrategy()
            .getDockerHostIpAddress()
            .shouldBeEqualTo(gateway)
```

The output when running on our gitlab (with support for docker-in-docker by passing the docker socket) is:

> Network: com.github.dockerjava.api.model.Network@4947a298[id=0742db67f2ff5b4706b8a3340248cd27e6d93c785304116e3c9b4911c1ea5a36
    ,name=bridge
    ,scope=local
    ,driver=bridge
    ,enableIPv6=false
    ,internal=false
    ,ipam=com.github.dockerjava.api.model.Network$Ipam@66efc28c[driver=default
    ,config=[com.github.dockerjava.api.model.Network$Ipam$Config@6ada782c]
    ,options=<null>]
    ,containers={b5de758269581432f830f7e78f8fe1e79ea5921eab554415f0c87102f25c90f9=com.github.dockerjava.api.model.Network$ContainerNetworkConfig@2fd37407[endpointId=afebcc4c3cf2e94b4309ccd2ee494ff73cdd9b7b0c4e44b360c93ef0f8084dc0
    ,macAddress=02:42:ac:11:00:02
    ,ipv4Address=172.17.0.2/16
    ,ipv6Address=]
    , edbef2cd0d08fca3f54e9149d0a3fdae91c76bafb1b8245bab101384ed450f7d=com.github.dockerjava.api.model.Network$ContainerNetworkConfig@5bf45007[endpointId=f20d40ccd6067864805411fd14c3d030f9be8fc8d5aed144b60c6540b9eed902
    ,macAddress=02:42:ac:11:00:03
    ,ipv4Address=172.17.0.3/16
    ,ipv6Address=]}
    ,options={com.docker.network.bridge.default_bridge=true
        , com.docker.network.bridge.enable_icc=true
        , com.docker.network.bridge.enable_ip_masquerade=true
        , com.docker.network.bridge.host_binding_ipv4=0.0.0.0
        , com.docker.network.bridge.name=docker0
        , com.docker.network.driver.mtu=1500}
        ,attachable=false
        ,labels={}]
Gateway: localhost
java.lang.AssertionError: Expected localhost actual 172.17.0.1.

Thanks for your help.

--- a/core/src/main/java/org/testcontainers/dockerclient/DockerClientProviderStrategy.java
+++ b/core/src/main/java/org/testcontainers/dockerclient/DockerClientProviderStrategy.java
@@ -218,7 +218,9 @@ public abstract class DockerClientProviderStrategy {
                         .filter(it -> it.getGateway() != null)
                         .findAny()
                         .map(Network.Ipam.Config::getGateway)
-                        .orElse("localhost");
+                        .orElseGet(() -> {
+                            return DockerClientConfigUtils.getDefaultGateway().orElse("localhost");
+                        });
                 }
                 return "localhost";
             default:

